atat_internal_api.yaml

```diff
@@ -0,0 +1,2197 @@
+    Application:
+      required:
+        - environments
+        - name
+        - operators
+      type: object
+      allOf:
+        - $ref: "#/components/schemas/BaseObject"
+      properties:
+        environments:
+          minItems: 1
+          type: array
+          items:
+            $ref: "#/components/schemas/Environment"
+        name:
+          pattern: "^[a-zA-Z\\d _-]{4,100}$"
+          type: string
+        description:
+          pattern: "^[\\w\\d !@#$%^&*_|:;,'.-]{0,300}$"
+          type: string
+      additionalProperties: false
+      description: "Represents an Application in a Portfolio"
+    ApplicationDetail:
+      allOf:
+        - $ref: '#/components/schemas/Application'
+        - $ref: '#/components/schemas/AppEnvAccess'
+    Environment:
+      required:
+        - name
+        - operators
+      type: object
+      allOf:
+        - $ref: "#/components/schemas/BaseObject"
+      properties:
+        name:
+          pattern: "^[a-zA-Z\\d _-]{1,100}$"
+          type: string
+      additionalProperties: false
+      description: "Represents an Environment for a specific Application"
+    EnvironmentDetail:
+      allOf:
+        - $ref: '#/components/schemas/Environment'
+        - $ref: '#/components/schemas/AppEnvAccess'
+    Clin:
+      required:
+        - clin_number
+        - idiq_clin
+        - obligatedFunds
+        - popEndDate
+        - popStartDate
+        - totalClinValue
+      type: object
+      allOf:
+        - $ref: "#/components/schemas/BaseObject"
+      properties:
+        clinNumber:
+          pattern: "(?!^0{4}$)^\\d{4}$"
+          type: string
+          description: "Contract Line Item Number (CLIN), 0001 through 9999"
+        obligatedFunds:
+          type: number
+        idiqClin:
+          type: string
+        popStartDate:
+          type: string
+          description: "Start of period of performance"
+          format: date
+        popEndDate:
+          type: string
+          description: "End of period of performance"
+          format: date
+        totalClinValue:
+          type: number
+      additionalProperties: false
+      description: "Represents a Clin in a Task Order"
+    TaskOrder:
+      required:
+        - clins
+        - taskOrderFile
+        - taskOrderNumber
+      type: object
+      allOf:
+        - $ref: "#/components/schemas/BaseObject"
+      properties:
+        clins:
+          minItems: 1
+          type: array
+          items:
+            $ref: "#/components/schemas/Clin"
+        taskOrderFile:
+          description: "Metadata associated with file which was previously uploaded\
+            \ by POST /taskOrderFiles"
+          allOf:
+            - $ref: "#/components/schemas/FileMetadataSummary"
+        taskOrderNumber:
+          pattern: "^[\\d]{13,17}$"
+          type: string
+          description: "TO numbers are 13 characters. TO modifications are 17 characters."
```

Comment:
I just want to point out that the business rules for this changed, and I believe there is tech debt to change the pattern to reflect it. Biggest change is making character 9 an F, but full list is on 2.1 of: https://ccpo.atlassian.net/wiki/spaces/AT/pages/2031747073/ATAT+Business+Rules+Data+Fields

Reply:
Some something more like

```regex
(?<DoDAAC>[a-z0-9]{6})
(?<year>\d{2})
F
(?<ToId>\d{4})
(?<ModId>\d{4})?
```

But probably without the named capture groups, so:

```
[a-z0-9]{6}\d{2}F\d{4}(?:\d{4})?
```

Technically that last group probably could be capturing as well (so remove `?:`) if we really wanted to simply it further. But it has to be grouped because `\d{4}?` means something different.

But is this a spec problem or a business logic problem?

Reply:
Agree the old pattern was based on limited business rules that have since been replaced.  Good note Zach.

Thanks for breaking down the new rules in a regular expression Kyle.

I think through his final question Kyle means that the purpose of this ticket/PR is new API spec, and yes, the old API spec did have the latent business rule problem.  Let's not address this here.  I've added a comment on business rules page to point to Kyle's regex comment for future reference when we do address this problem.